### PR TITLE
copr: batch handle request

### PR DESCRIPTION
**Pull request by BusyJay**
@ngaut @siddontang @disksing PTAL

### CODE REVIEW

File: src/server/server.rs [hunk 1]
```diff
@@ -28,15 +28,17 @@ use super::{Msg, SendCh, ConnData};
...
 use super::kv::StoreHandler;
-use super::coprocessor::EndPointHost;
+use super::coprocessor::{RequestTask, EndPointHost};
 use super::transport::RaftStoreRouter;
 use super::resolve::StoreAddrResolver;
 use raft::SnapshotStatus;
 
 const SERVER_TOKEN: Token = Token(1);
 const FIRST_CUSTOM_TOKEN: Token = Token(1024);
+const DEFAULT_COPROCESSOR_CONCURRENCY: usize = 50;
```
Comment (siddontang) [inline src/server/server.rs]: Concurrency confuses me here, do we have 50 threads to do coprocessor? Seem batch_size here.

File: src/server/coprocessor.rs [hunk 2]
```diff
@@ -106,15 +108,46 @@ impl EndPointHost {
...
 
-    pub fn on_request(&self, req: Request, on_resp: OnResponse) {
-        let end_point = self.snap_endpoint.clone();
-        self.pool.execute(move || {
-            let timer = SlowTimer::new();
-            let tp = req.get_tp();
-            end_point.handle_request(req, on_resp);
-            slow_log!(timer, "request type {} takes {:?}", tp, timer.elapsed());
-        });
+pub struct RequestTask(Request, OnResponse);
```
Comment (siddontang) [inline src/server/coprocessor.rs]: I prefer name field structure. not tuple.

File: src/server/coprocessor.rs [hunk 3]
```diff
@@ -106,15 +108,46 @@ impl EndPointHost {
...
+    fn run_batch(&mut self, reqs: &mut Vec<RequestTask>) {
+        let mut grouped_reqs = map![];
+        for req in reqs.drain(..) {
+            let key = {
+                let ctx = req.0.get_context();
+                (ctx.get_region_id(),
+                 ctx.get_region_epoch().get_conf_ver(),
+                 ctx.get_region_epoch().get_version(),
+                 ctx.get_peer().get_id(),
+                 ctx.get_peer().get_store_id())
```
Comment (siddontang) [inline src/server/coprocessor.rs]: do we need store id? it is can't changed. 

can we use region id only?

Comment (BusyJay) [inline src/server/coprocessor.rs]: What if `store_id` is wrong? Only `region_id` is not enough, may contains stale epoch.

Comment (siddontang) [inline src/server/coprocessor.rs]: don't we check store id and stale epoch when executing the request? 
if wrong, or stale epoch, I think we can return error response.

Comment (BusyJay) [inline src/server/coprocessor.rs]: Yes, but it's not executed yet. Here just group the requests before dispatching.

**Comment by disksing** [comment]
LGTM.

File: tests/coprocessor/test_select.rs [hunk 4]
```diff
@@ -171,7 +193,7 @@ fn prepare_sel(store: &mut Store, tbl: &TableInfo) -> Request {
     req
 }
 
-fn initial_data(count: i64) -> (Store, TiDbEndPoint, TableInfo) {
+fn initial_data(count: i64) -> (Store, Worker<RequestTask>, TableInfo) {
     let engine = Arc::new(engine::new_engine(Dsn::RocksDBPath(TEMP_DIR)).unwrap());
     let mut store = Store::new(engine.clone());
     let tbl = TableInfo {
```
Comment (ngaut) [inline tests/coprocessor/test_select.rs]: s/tbl/ti/

**Comment by ngaut** [comment]
LGTM

**Comment by siddontang** [comment]
LGTM

**Pull request merged**